libs/providers/flagd-web/src/lib/flagd-web-provider.ts

```diff
@@ -0,0 +1,240 @@
+  private _retry = 0;
+  private _delayMs = INITIAL_DELAY_MS;
+  private _streamingEnabled: boolean;
+  private _logger?: Logger;
+
+  constructor(
+    options?: FlagdProviderOptions,
+    logger?: Logger,
+    cache?: FlagdCache,
+    promiseClient?: PromiseClient<typeof Service>,
+    callbackClient?: CallbackClient<typeof Service>,
+  ) {
+    const { host, port, tls, caching, cacheTtl, maxRetries, maxDelay, eventStreaming } = getOptions(options);
+    const transport = createConnectTransport({
+      baseUrl: `${tls ? 'https' : 'http'}://${host}:${port}`,
+    });
+    this._promiseClient = promiseClient ? promiseClient : createPromiseClient(Service, transport);
+    this._callbackClient = callbackClient ? callbackClient : createCallbackClient(Service, transport);
+    this._cache = cache ? cache : new LocalStorageFlagdCache({ itemTtl: cacheTtl }, logger);
+    this._maxRetries = maxRetries === 0 ? Infinity: maxRetries;
+    this._maxDelay = maxDelay;
+    this._streamingEnabled = eventStreaming;
+    this._cacheEnabled = caching;
+    this._logger = logger;
+
+    if (this._streamingEnabled) {
+      this.retryConnect();
+    } else {
+      this._connected = true;
+    }
+  }
+
+  // if the cache is active, or we're connected, or streaming is disabled, we're ready.
+  get ready() {
+    return this._connected || this._cacheEnabled || !this._streamingEnabled;
+  }
+
+  resolveBooleanEvaluation(
+    flagKey: string,
+    defaultValue: boolean,
+    context: EvaluationContext
+  ): Promise<ResolutionDetails<boolean>> {
+    return this.evaluate(flagKey, defaultValue, context, this._promiseClient.resolveBoolean);
+  }
+
+  resolveStringEvaluation(
+    flagKey: string,
+    defaultValue: string,
+    context: EvaluationContext
+  ): Promise<ResolutionDetails<string>> {
+    return this.evaluate(flagKey, defaultValue, context, this._promiseClient.resolveString);
+  }
+
+  resolveNumberEvaluation(
+    flagKey: string,
+    defaultValue: number,
+    context: EvaluationContext
+  ): Promise<ResolutionDetails<number>> {
+    return this.evaluate(flagKey, defaultValue, context, this._promiseClient.resolveFloat);
+  }
+
+  resolveObjectEvaluation<U extends JsonValue>(
+    flagKey: string,
+    defaultValue: U,
+    context: EvaluationContext
+  ): Promise<ResolutionDetails<U>> {
+    return this.evaluate(flagKey, defaultValue, context, this._promiseClient.resolveObject);
+  }
+
+  private async evaluate<T extends FlagValue>(
+    flagKey: string,
+    defaultValue: T,
+    context: EvaluationContext,
+    resolver: (
+      request: PartialMessage<ResolveAnyRequest>,
+      options?: CallOptions | undefined
+    ) => Promise<ResolveAnyResponse>
+  ): Promise<ResolutionDetails<T>> {
+    try {
+      const result = await this._cache.get<T>(flagKey, context);
+      if (this._cacheEnabled && result) {
+        return result;
+      }
+      const res = await resolver({flagKey, context: Struct.fromJson(context as JsonValue)});
```

Comment:
Because the type is functionally the same, but not exactly - one is from our SDK and one is from the grpc lib.

Before (and in flagd) this is sidestep by doing `Struct.fromJsonString(JSON.stringify(context))` but that's less performant, a just hides the exact same issue from the compiler.